core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -371,6 +371,14 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
       throw new SparkException("An application name must be set in your configuration")
     }
 
+    // System property spark.yarn.app.id must be set if user code ran by AM on a YARN cluster
+    // yarn-standalone is deprecated, but still supported
+    if ((master == "yarn-cluster" || master == "yarn-standalone") &&
+        !_conf.contains("spark.yarn.app.id")) {
```

Comment:
yea, the suggestion was though that this should also be `spark.app.id`, since the two usages of `spark.yarn.app.id` are the only two in the code base: https://github.com/apache/spark/pull/6120  It might be worth resolving that first since it will matter to this change.